### PR TITLE
implements svg generation

### PR DESCRIPTION
**Pull request by Grummfy**

### CODE REVIEW

File: src/Identicon/Generator/BaseGenerator.php [hunk 1]
```diff
@@ -238,4 +179,63 @@ public function getPixelRatio()
...
+			if (false !== strpos($color, '#')) {
+				$color = substr($color, 1);
+			}
+			$convertedColor[0] = hexdec(substr($color, 0, 2));
+			$convertedColor[1] = hexdec(substr($color, 2, 2));
+			$convertedColor[2] = hexdec(substr($color, 4, 2));
+		}
+
+		return $convertedColor;
+	}
```
Comment (blaugueux) [inline src/Identicon/Generator/BaseGenerator.php]: Why did you change the indentation in this file?

Comment (blaugueux) [inline src/Identicon/Generator/BaseGenerator.php]: Please fix the global indentation. You need to use the 4 spaces indentation.

File: src/Identicon/Generator/SvgGenerator.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
+<?php
+
+namespace Identicon\Generator;
+
+use Identicon\Generator\GeneratorInterface;
+
+/**
+ * @author Benjamin Laugueux <benjamin@yzalis.com>
```
Comment (blaugueux) [inline src/Identicon/Generator/SvgGenerator.php]: Put your name here.

File: src/Identicon/Identicon.php [hunk 3]
```diff
@@ -95,6 +95,6 @@ public function getImageResource($string, $size = 64, $color = null, $background
      */
     public function getImageDataUri($string, $size = 64, $color = null, $backgroundColor = null)
     {
-        return sprintf('data:image/png;base64,%s', base64_encode($this->getImageData($string, $size, $color, $backgroundColor)));
+        return sprintf('data:' . $this->generator->getMimeType() . ';base64,%s', base64_encode($this->getImageData($string, $size, $color, $backgroundColor)));
```
Comment (blaugueux) [inline src/Identicon/Identicon.php]: Add `%s` and your replacement as an argument.

File: src/Identicon/Generator/SvgGenerator.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+		$backgroundColor = '#FFFFFF';
+		$rgbBackgroundColor = $this->getBackgroundColor();
+		if (!is_null($rgbBackgroundColor))
+		{
+			$backgroundColor = $this->_toUnderstandableColor($rgbBackgroundColor);
+		}
+		$svg .= '<rect width="'. $w .'" height="'. $h .'" style="fill:'. $backgroundColor .';stroke-width:1;stroke:' . $backgroundColor .'"/>';
+
+
```
Comment (blaugueux) [inline src/Identicon/Generator/SvgGenerator.php]: Remove blank line.

File: src/Identicon/Generator/SvgGenerator.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	{
+		// prepare image
+		$w = $this->getPixelRatio() * 5;
+		$h = $this->getPixelRatio() * 5;
+		$svg = '<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="'. $w .'" height="'. $h .'">';
+
+		$backgroundColor = '#FFFFFF';
+		$rgbBackgroundColor = $this->getBackgroundColor();
+		if (!is_null($rgbBackgroundColor))
+		{
```
Comment (blaugueux) [inline src/Identicon/Generator/SvgGenerator.php]: Bracket should be at the end of the previous line.

File: src/Identicon/Generator/SvgGenerator.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+			$backgroundColor = $this->_toUnderstandableColor($rgbBackgroundColor);
+		}
+		$svg .= '<rect width="'. $w .'" height="'. $h .'" style="fill:'. $backgroundColor .';stroke-width:1;stroke:' . $backgroundColor .'"/>';
+
+
+		$rgbColor = $this->_toUnderstandableColor($this->getColor());
+		// draw content
+		foreach ($this->getArrayOfSquare() as $lineKey => $lineValue)
+		{
```
Comment (blaugueux) [inline src/Identicon/Generator/SvgGenerator.php]: Bracket should be at the end of the previous line.

File: src/Identicon/Generator/SvgGenerator.php [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+			foreach ($lineValue as $colKey => $colValue)
+			{
```
Comment (blaugueux) [inline src/Identicon/Generator/SvgGenerator.php]: Bracket should be at the end of the previous line.

File: src/Identicon/Generator/SvgGenerator.php [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+				if (true === $colValue)
+				{
```
Comment (blaugueux) [inline src/Identicon/Generator/SvgGenerator.php]: Bracket should be at the end of the previous line.

File: src/Identicon/Generator/SvgGenerator.php [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+	}
+
+	/**
+	 * @param array|string $color
+	 * @return string
+	 */
+	protected function _toUnderstandableColor($color)
+	{
+		if (is_array($color))
+		{
```
Comment (blaugueux) [inline src/Identicon/Generator/SvgGenerator.php]: Bracket should be at the end of the previous line.

File: src/Identicon/Generator/SvgGenerator.php [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+		$svg .= '</svg>';
+
+		$this->generatedImage = $svg;
+
+		return $this;
+	}
+
+	/**
+	 * @param array|string $color
+	 * @return string
```
Comment (blaugueux) [inline src/Identicon/Generator/SvgGenerator.php]: Add a blank line before `@return`

File: src/Identicon/Generator/GeneratorInterface.php [hunk 11]
```diff
@@ -29,5 +25,11 @@ function getImageBinaryData($string, $size = null, $color = null, $backgroundCol
      *
      * @return string
      */
-    function getImageResource($string, $size = null, $color = null, $backgroundColor = null);
+    public function getImageResource($string, $size = null, $color = null, $backgroundColor = null);
+
+	/**
+	 * Return the mime-type of this identicon
+	 * @return string
```
Comment (blaugueux) [inline src/Identicon/Generator/GeneratorInterface.php]: Add a blank line before `@return`

**Comment by blaugueux** [comment]
Nice feature. A few changes and i'll merge it!

**Comment by Grummfy** [comment]
Can you give me the coding standard (or better a phpcs file rule) you use to adapt the code as you like?

**Comment by blaugueux** [comment]
Sorry but I don't have time to write a phpcs file. Please just check the existing code. Thx

**Comment by blaugueux** [comment]
Could you also add some lines in the Readme so all devs can use your feature easily.

**Comment by Grummfy** [comment]
here we are, all you should be fixed.

File: src/Identicon/Generator/BaseGenerator.php [hunk 12]
```diff
@@ -238,4 +179,63 @@ public function getPixelRatio()
...
+            if (false !== strpos($color, '#')) {
+                $color = substr($color, 1);
+            }
+            $convertedColor[0] = hexdec(substr($color, 0, 2));
+            $convertedColor[1] = hexdec(substr($color, 2, 2));
+            $convertedColor[2] = hexdec(substr($color, 4, 2));
+        }
+
+        return $convertedColor;
+    }
```
Comment (blaugueux) [inline src/Identicon/Generator/BaseGenerator.php]: Why this file has some changes?

Comment (blaugueux) [inline src/Identicon/Generator/BaseGenerator.php]: It seems that you have just switch methods.

File: src/Identicon/Generator/GeneratorInterface.php [hunk 13]
```diff
@@ -17,11 +15,9 @@
      *
      * @return mixed
      */
-    function getImageBinaryData($string, $size = null, $color = null, $backgroundColor = null);
+    public function getImageBinaryData($string, $size = null, $color = null, $backgroundColor = null);
```
Comment (blaugueux) [inline src/Identicon/Generator/GeneratorInterface.php]: Good catch.

File: src/Identicon/Generator/GeneratorInterface.php [hunk 14]
```diff
@@ -29,5 +25,12 @@ function getImageBinaryData($string, $size = null, $color = null, $backgroundCol
      *
      * @return string
      */
-    function getImageResource($string, $size = null, $color = null, $backgroundColor = null);
+    public function getImageResource($string, $size = null, $color = null, $backgroundColor = null);
```
Comment (blaugueux) [inline src/Identicon/Generator/GeneratorInterface.php]: Good catch.

File: src/Identicon/Generator/BaseImagePngGenerator.php [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * @author Grummfy <grummfy@gmail.com>
+ */
+class BaseImagePngGenerator extends BaseGenerator
+{
+	public function getMimeType()
+	{
+		return 'image/png';
+	}
+}
```
Comment (blaugueux) [inline src/Identicon/Generator/BaseImagePngGenerator.php]: Maybe we should add the mime type inside each generator (Gd, Imagick, Svg). This will avoid an other "Base" method.

File: src/Identicon/Generator/BaseGenerator.php (hunk 12 above)
Comment (Grummfy) [inline src/Identicon/Generator/BaseGenerator.php]: Normally I just fix the identation, but perhaps (I use auto fix of phpstorm) some other change have been made. Sorry for that. I can go back and make a rebase to assemble each commit to one, to have a more clear vision of all change.

**Comment by blaugueux** [comment]
ping @Grummfy. Do you have time to finish this PR? Thanks for your work on this.

**Comment by Grummfy** [comment]
oups completely forgot .... will add this to my agenda ...

**Comment by Grummfy** [comment]
I think it's finsihed, everything is woking well.

**Comment by blaugueux** [comment]
I don't see any new code :(

**Comment by Grummfy** [comment]
yes but What do you want more, it's working ...

File: src/Identicon/Generator/BaseGenerator.php (hunk 12 above)
Comment (blaugueux) [inline src/Identicon/Generator/BaseGenerator.php]: Please revert this file if there is no change.

File: src/Identicon/Generator/BaseImagePngGenerator.php (hunk 15 above)
Comment (blaugueux) [inline src/Identicon/Generator/BaseImagePngGenerator.php]: This method must be inside each Generator. We do not need a ```BaseImagePngGenerator``` class.

**Comment by blaugueux** [comment]
You also need to squash your commits so I can merge your PR. Thx

**Comment by Grummfy** [comment]
oki doki

**Comment by Grummfy** [comment]
here we are. It should be good now

File: src/Identicon/Generator/BaseGenerator.php [hunk 16]
```diff
@@ -7,7 +7,7 @@
 /**
  * @author Benjamin Laugueux <benjamin@yzalis.com>
  */
-class BaseGenerator
+abstract class BaseGenerator
```
Comment (blaugueux) [inline src/Identicon/Generator/BaseGenerator.php]: Moving this class to astract cause test to fail.

File: src/Identicon/Generator/BaseGenerator.php [hunk 17]
```diff
@@ -44,6 +44,8 @@ class BaseGenerator
      */
     private $arrayOfSquare = array();
 
+    abstract public function getMimeType();
```
Comment (blaugueux) [inline src/Identicon/Generator/BaseGenerator.php]: This method is already inside the Interface. I think we do not need this class to be abstract.

**Comment by blaugueux** [comment]
Almost done @Grummfy !

**Comment by Grummfy** [comment]
Because locally (php 5.6) test failed, I add some extra for travis (5.6 & 7.0).

**Comment by blaugueux** [comment]
Good job!

**Comment by Grummfy** [comment]
cool ;)
